### DISCUSSION TITLE
New dev tool - Burocrata: Check and insert copyright and license notices into source code

### DISCUSSION THREAD

**Discussion by leouieda** [fatiando/community#139] [Announcements]
## About

For a while now, we've had a [Python script](https://github.com/fatiando/verde/blob/05b169c5b145718a0ea25ed175636ae43f39f745/tools/license_notice.py) that @santisoler wrote to check if all our source code files start with a copyright and license notice ([like this one](https://github.com/fatiando/verde/blob/05b169c5b145718a0ea25ed175636ae43f39f745/verde/blockreduce.py#L1). This script was copied to every repository in our organization and was used to make sure we don't forget to add the notice to new source files we create.

Found myself wanting to code something yesterday and finally got around to building a proper command-line (CLI) app for doing this. **Burocrata** is a little Python app that can check if the license notice is in the source files and optionally add it to any missing files. Our previous script hard-coded the license notice but Burocrata can read it from the `pyproject.toml` files.

## Using

Check that very `.py` file in a directory has a license notice:

```
$ burocrata --check --extension=py source_folder
```

Removing the `--check` option will make Burocrata add the license notice to
the files that don't have them:

```
$ burocrata --extension=py source_folder
```

The license and copyright notice can be configured in a `pyproject.toml` file
located in the directory where `burocrata` is run:

```
$ cat pyproject.toml
[tool.burocrata]
notice = '''
# Copyright (c) YYYY Name of Developer.
# Distributed under the terms of the BSD 3-Clause License.
# SPDX-License-Identifier: BSD-3-Clause'''
```

## Installing

*Burocrata* is available from PyPI:

```
python -m pip install burocrata
```

A conda-forge package will be coming soon.

## What do you think?

Feedback and code contributions are more than welcome!

**Comment by santisoler** [comment]
Love to see this! I write a little script and @leouieda turns it into a full CLI with `pyproject.toml` support ❤️ 

I'm leaving the link to Burocrata's repo (that I didn't see in your post): https://github.com/fatiando/burocrata

Just a few ideas (nothing urgent):
- Can we pass a single file instead a whole folder?
- Could we include an `exclude` option in the `pyproject.toml` file to ignore particular files?

**Reply by leouieda** [reply]
Thanks, @santisoler! I basically copied the CLI from dependente and had your script for most of the rest so this was quite easy.

> Could we include an exclude option in the pyproject.toml file to ignore particular files?

It will ignore whatever is on the `.gitignore` but if you want to ignore other files then we could add a list to `pyproject.toml` with the same syntax as `.gitignore` and use the same parser to exclude the files. Would you mind opening an issue?

> Can we pass a single file instead a whole folder?

Sure. I guess this would be the case of adding an `if path.is_file()` check before trying to glob the folder. Issue 😇?

**Reply by santisoler** [reply]
Sure. I'll open those issues, just wanted to check first if this is something we would like to have. Thanks again for the energy put into this!
